service/frontend/workflowHandler.go

```diff
@@ -3225,3 +3267,201 @@ func (wh *WorkflowHandler) getArchivedHistory(
 		Archived:      common.BoolPtr(true),
 	}, nil
 }
+
+func (wh *WorkflowHandler) convertIndexedKeys(keys map[string]interface{}) map[string]gen.IndexedValueType {
+	converted := make(map[string]gen.IndexedValueType)
+	for k, v := range keys {
+		switch v {
+		case gen.IndexedValueTypeString:
+			converted[k] = gen.IndexedValueTypeString
+		case gen.IndexedValueTypeKeyword:
+			converted[k] = gen.IndexedValueTypeKeyword
+		case gen.IndexedValueTypeInt:
+			converted[k] = gen.IndexedValueTypeInt
+		case gen.IndexedValueTypeDouble:
+			converted[k] = gen.IndexedValueTypeDouble
+		case gen.IndexedValueTypeBool:
+			converted[k] = gen.IndexedValueTypeBool
+		case gen.IndexedValueTypeDatetime:
+			converted[k] = gen.IndexedValueTypeDatetime
+		default:
+			wh.GetLogger().Fatal("unknown index value type", tag.Value(v))
+		}
+	}
+	return converted
+}
+
+func (wh *WorkflowHandler) validateSearchAttributes(input *gen.SearchAttributes, domain string) error {
+	if input == nil {
+		return nil
+	}
+
+	fields := input.GetIndexedFields()
+	lengthOfFields := len(fields)
+	if lengthOfFields > wh.config.SearchAttributesNumberOfKeysLimit(domain) {
+		wh.GetLogger().WithTags(tag.Number(int64(lengthOfFields)), tag.WorkflowDomainName(domain)).
+			Error("number of keys in search attributes exceed limit")
+		return fmt.Errorf("number of keys %d exceed limit", lengthOfFields)
+	}
+
+	totalSize := 0
+	for key, val := range fields {
+		if !wh.isValidSearchAttributes(key) {
+			wh.GetLogger().WithTags(tag.ESKey(key), tag.WorkflowDomainName(domain)).
+				Error("invalid search attribute")
+			return fmt.Errorf("%s is not valid search attribute", key)
+		}
+		if definition.IsSystemIndexedKey(key) {
+			wh.GetLogger().WithTags(tag.ESKey(key), tag.WorkflowDomainName(domain)).
+				Error("illegal update of system reserved attribute")
+			return fmt.Errorf("%s is read-only Cadence reservered attribute", key)
+		}
+		if len(val) > wh.config.SearchAttributesSizeOfValueLimit(domain) {
+			wh.GetLogger().WithTags(tag.ESKey(key), tag.Number(int64(len(val))), tag.WorkflowDomainName(domain)).
+				Error("value size of search attribute exceed limit")
+			return fmt.Errorf("size limit exceed for key %s", key)
+		}
+		totalSize += len(key) + len(val)
+	}
+
+	if totalSize > wh.config.SearchAttributesTotalSizeLimit(domain) {
+		wh.GetLogger().WithTags(tag.Number(int64(totalSize)), tag.WorkflowDomainName(domain)).
+			Error("total size of search attributes exceed limit")
+		return fmt.Errorf("total size %d exceed limit", totalSize)
+	}
+
+	return nil
+}
+
+// validateListRequestForQuery validate that search attributes in listRequest query is legal,
+// and add prefix for custom keys
+func (wh *WorkflowHandler) validateListRequestForQuery(listRequest *gen.ListWorkflowExecutionsRequest) error {
```

Comment:
Can you have a separate QueryValidator component instead of having separate functions for individual validation defined as functions.

Reply:
Updated